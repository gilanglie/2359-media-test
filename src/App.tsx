import React from 'react';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Box from '@material-ui/core/Box';
import Paper from '@material-ui/core/Paper';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
import LinearProgress from '@material-ui/core/LinearProgress';
import CircularProgress from '@material-ui/core/CircularProgress';
import Typography from '@material-ui/core/Typography';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Button from '@material-ui/core/Button';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      width: '100%',
      '& > * + *': {
        marginTop: theme.spacing(2),
      },
      
    },
    input:{
      display: 'none',
    },
    title: {
      flexGrow: 1,
    },
  }),
);
export default function App(props:any) {
  const classes = useStyles();

  const [limit, setLimit] = React.useState(0);
  const [bars, setBars] = React.useState([]);
  const [buttons, setButtons] = React.useState([]);
  const [selected, setSelected] = React.useState(0);
  async function fetchData() {
    try{
      const res = await fetch("https://pb-api.herokuapp.com/bars")
      const json = await res.json();
      setButtons(json.buttons);
      setLimit(json.limit);
      const barData = json.bars.map((option:any)=>{
        const childData = {
          progress : option,
          elevation : 1
        }
        return childData;
      })
      barData[0].elevation = 4;
      setBars(barData);
      
    }catch(err){
      console.log(err);
    }
  }

  React.useEffect(() => {
      fetchData();
  }, []);
  const handleRadio = (event:any) => {
    const items: any = bars.map((option:any,index:any)=>{
      if(index == event.target.value){
        return {
          progress : option.progress,
          elevation : 4
        }
      }else{
        return {
          progress : option.progress,
          elevation : 1
        }
      }
    });
    setBars(items);
    setSelected(event.target.value);
  }
  const handleClick = (event:any,val:number) =>{
    const barData:any = [...bars]
    barData[selected].progress =  barData[selected].progress + val;
    setBars(barData);
  }
  return (
    <div className={classes.root}>
      <AppBar position="static">
        <Toolbar>
          <Grid container alignItems="center">
            <Grid item xs={12} sm={6}>
              <Typography variant="h6" className={`${classes.title} header-title`}>
                Progress Bar Demo
              </Typography>
              <Typography variant="caption">Limit : {limit}</Typography>
            </Grid>
            <Grid item xs={12} sm={6}>
                {buttons.map((option:any)=>(
                  <Button  color="inherit" onClick={(e:any) => handleClick(e,option)}>{option}</Button>
                ))}

            </Grid>
          </Grid>
          

        </Toolbar>
      </AppBar>
      <Container>
        
        <Grid container spacing={4} alignItems="center" justify="center">
          {bars.length == 0 ? 
            <Grid item className="loading">
              <CircularProgress size={100}/>
              <Typography variant="h6">Fetching Data...</Typography>
            </Grid>
          :
            bars.map((option:any, index:any)=>(
              <Grid item sm={6} xs={12}>
              <label>
              <input type="radio"  
              name="progressBar" 
              className={classes.input}
              value={index}
              onChange={handleRadio}
              />
                <Paper elevation={option.elevation}>
                  <Box p={4} m={2}  >
                    <LinearProgress variant="determinate" value={option.progress} color={option.progress > limit ? "secondary" : "primary"}/>
                    <Typography variant="subtitle1">
                      {option.progress}%
                    </Typography>
                  </Box>
                </Paper>
              </label>
            </Grid>
            ))
          }
          </Grid>
      </Container>
    </div>
  );
} 
