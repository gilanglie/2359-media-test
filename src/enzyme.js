import React from 'react';
import Enzyme, { configure, shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
global.React = React;
global.shallow = shallow;
global.render = render;
global.mount = mount;

export { shallow, mount, render };
export default Enzyme;
