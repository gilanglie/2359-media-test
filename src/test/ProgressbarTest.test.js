import React from 'react';
import { shallow, mount, render } from '../enzyme';
import { TextField, Button, Grid } from '@material-ui/core';
import "whatwg-fetch";
import fetchMock from "fetch-mock";
import { act } from "react-test-renderer";
import renderer from 'react-test-renderer';
import { renderHook } from "@testing-library/react-hooks";

import App from '../App';

let wrapper, props, useEffect;
const mockUseEffect = () => {
    useEffect.mockImplementationOnce(f => f());
  };
  const posts = [{ id: 1, title: "a post", body: "the body" }];

beforeEach(() => {
    useEffect = jest.spyOn(React, "useEffect");

    props = {
        fetchData: jest.fn().mockResolvedValue(posts)
    };
    mockUseEffect();
    wrapper = shallow(<App {...props} /> );
});


describe('ProgressBar Test', () => {
    beforeAll(() => {
        global.fetch = fetch;
    });
    afterAll(() => {
        jest.clearAllMocks();
    });
    
    it('matches the snapshot', () => {
        const tree = renderer.create(<App />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("rendered", () => {
        shallow(<App />);
    });

    it("displays loading", () => {
        expect(wrapper.exists(".loading"));
    });

    it('should render 2 <Grid />', () => {
        expect(wrapper.find(Grid)).toHaveLength > 0;
    });
    // it('should render 2 <Button />', () => {
    //     const setState = jest.fn();
    //     const useStateSpy = jest.spyOn(React, 'useState')
    //     expect(props.fetchData).toHaveBeenCalled();

    //     useStateSpy.mockImplementation((init) => [init, setState]);
    //     expect(setState).toHaveLength > 0;
    //     const tree = renderer.create(<App />).toJSON();
    //     expect(tree).toMatchSnapshot();
    // });
});

describe('<Button /> interactions', () => {
    it('should call the onClick function whenbutton is clicked', () => {
        expect(wrapper.find(Button).onClick);
    });
})
      

