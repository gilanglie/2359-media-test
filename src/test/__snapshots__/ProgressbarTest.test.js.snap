// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar Test matches the snapshot 1`] = `
<div
  className="makeStyles-root-1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <h6
        className="MuiTypography-root makeStyles-title-3 header-title MuiTypography-h6"
      >
        Progress Bar Demo
      </h6>
    </div>
  </header>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      <div
        className="MuiGrid-root loading MuiGrid-item"
      >
        <div
          className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style={
            Object {
              "height": 100,
              "width": 100,
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={Object {}}
            />
          </svg>
        </div>
        <h6
          className="MuiTypography-root MuiTypography-h6"
        >
          Fetching Data...
        </h6>
      </div>
    </div>
  </div>
</div>
`;
